File: Algorithms/Flow/Algorithms/Part_1_Simple_Algorithms/3.Generating permutations/GeneratingPermutations.swift
//
//  GeneratingPermutations.swift
//  Algorithms
//
//  Created by Maksim Ponomarev on 12/6/24.
//

import Foundation

func generatePermutations(
	array: inout [Int],
	index: Int
) -> [[Int]] {
	// Array to hold all the permutations
	var results = [[Int]]()

	if index == array.count {
		// Add the current permutation to the results
		results.append(array)
		return results
	}
	
	for i in index..<array.count {
		// Swap the current index with the loop index
		array.swapAt(i, index)
		
		// Recur to generate permutations for the next index
		results += generatePermutations(
			array: &array,
			index: index + 1
		)
		
		// Backtrack by swapping back
		array.swapAt(i, index)
	}
	
	return results
}

File: Algorithms/Flow/Algorithms/Part_1_Simple_Algorithms/3.Generating permutations/Generating_permutations.cpp
//
//  Generating_permutations.cpp
//  Algorithms
//
//  Created by Maksim Ponomarev on 12/6/24.
//

#include <iostream>
#include <vector>

void generatePermutations(
						  std::vector<int>& nums,
						  int start
						  )
{
	// If we reach the end of the array, we have a complete permutation
	if (start == nums.size()) {
		// Print the current permutation
		std::cout << "{ ";
		for (int num : nums) {
			std::cout << num << " ";
		}
		std::cout << "}" << std::endl;
		return;
	}

	for (int i = start; i < nums.size(); ++i) {
		// Swap the current element with the start element
		std::swap(nums[start], nums[i]);
		
		// Recur with the next element as the starting point
		generatePermutations(nums, start + 1);
		
		// Backtrack by swapping back
		std::swap(nums[start], nums[i]);
	}
}

int main() {
	std::vector<int> nums = {1, 2, 3}; // Input set
	generatePermutations(nums, 0); // Start generating permutations

	return 0;
}
